dandi/metadata.py

```diff
@@ -83,3 +88,161 @@ def get_metadata(path):
+        duration = timedelta2duration(start - dob)
+    return models.PropertyValue(value=duration, unitText="Years from birth")
+
+
+def timedelta2duration(delta):
+    """
+    Convert a datetime.timedelta to ISO 8601 duration format
+
+    Parameters
+    ----------
+    delta : datetime.timedelta
+
+    Returns
+    -------
+    str
+    """
+    s = "P"
+    if delta.days:
+        s += f"{delta.days}D"
+    if delta.seconds or delta.microseconds:
+        sec = delta.seconds
+        if delta.microseconds:
+            # Don't add when microseconds is 0, so that sec will be an int then
+            sec += delta.microseconds / 1000000
+        s += f"T{sec}S"
+    if s == "P":
+        s += "0D"
+    return s
+
+
+def extract_sex(metadata):
+    if metadata.get("sex") is not None:
+        return models.SexType(identifier="sex", name=metadata["sex"])
+    else:
+        return ...
+
+
+def extract_assay_type(metadata):
+    if "assayType" in metadata:
+        return [models.AssayType(identifier="assayType", name=metadata["assayType"])]
+    else:
+        return []
+
+
+def extract_anatomy(metadata):
+    if "anatomy" in metadata:
+        return [models.Anatomy(identifier="anatomy", name=metadata["anatomy"])]
+    else:
+        return []
+
+
+def extract_model(modelcls, metadata, **kwargs):
+    m = modelcls.unvalidated()
+    for field in m.__fields__.keys():
+        value = kwargs.get(field, extract_field(field, metadata))
+        if value is not Ellipsis:
+            setattr(m, field, value)
+    # return modelcls(**m.dict())
+    return m
+
+
+def extract_wasDerivedFrom(metadata):
+    return [
+        extract_model(models.BioSample, metadata, identifier=metadata.get("subject_id"))
+    ]
+
+
+def extract_digest(metadata):
+    if "digest" in metadata:
+        return models.Digest(
+            value=metadata["digest"],
+            cryptoType=models.DigestType[metadata["digest_type"]],
+        )
+    else:
+        return ...
+
+
+FIELD_EXTRACTORS = {
+    "wasDerivedFrom": extract_wasDerivedFrom,
+    "age": extract_age,
+    "sex": extract_sex,
+    "assayType": extract_assay_type,
+    "anatomy": extract_anatomy,
+    "digest": extract_digest,
+}
+
+
+def extract_field(field, metadata):
+    if field in FIELD_EXTRACTORS:
+        return FIELD_EXTRACTORS[field](metadata)
+    else:
+        return metadata.get(field, ...)
+
+
+def nwb2asset(nwb_path, digest=None, digest_type=None):
+    start_time = datetime.now().astimezone()
+    metadata = get_metadata(nwb_path)
+    if digest is not None:
+        metadata["digest"] = digest
+        metadata["digest_type"] = digest_type
+    metadata["contentSize"] = op.getsize(nwb_path)
+    metadata["encodingFormat"] = "application/x-nwb"
+    asset = metadata2asset(metadata)
+    end_time = datetime.now().astimezone()
+    asset.wasGeneratedBy = models.Activity(
+        identifier=str(uuid4()),
+        name="NWB conversion",
+        description="Data converted using DANDI cli",
```

Comment:
just a note that `startedAt` and `endedAt` should reflect the creation of the asset. which may be a little more complicated than this. this `startedAt` and `endedAt` seems to be more closely associated with metadata creation.

```suggestion
        description="Metadata generated by DANDI cli",
```

since the description is about the asset and it may have multiple activities that generate the asset (conversion, organization) or its metadata (this function).

for now this is ok, but we may need to get more detailed with `isPartOf` and `hasPart` keywords.

Reply:
I think we might need to make generated by into a list. Current activity about metadata, so time stamps are for time when converted by dandi cli. But we might want to add activity on nwb generation/convertion - that one we might later deduce how it was generated etc